src/screen/dashboard.rs

```diff
@@ -128,105 +128,177 @@ impl Dashboard {
+                                                            server, nick, file,
+                                                        )
+                                                    },
+                                                ),
+                                                None,
+                                            );
+                                        }
+                                    }
                                 }
                             }
+                            Some(buffer::Event::ScrolledToTop) => {
+                                if config.buffer.chathistory.infinite_scroll {
+                                    if let Some(buffer) = pane.buffer.data() {
+                                        let server = buffer.server();
+
+                                        if clients.get_server_supports_chathistory(server) {
+                                            if let Some(target) = buffer.target() {
+                                                let message_reference_type = clients
+                                                    .get_server_chathistory_message_reference_type(
+                                                        server,
+                                                    );
+
+                                                let oldest_message_reference = self
+                                                    .get_oldest_message_reference(
+                                                        server,
+                                                        &target,
+                                                        message_reference_type,
+                                                    );
+
+                                                clients.send_chathistory_request(
+                                                    ChatHistorySubcommand::Before,
+                                                    server,
+                                                    &target,
+                                                    oldest_message_reference,
+                                                );
+                                            }
+                                        }
+                                    }
+                                }
+                            }
+                            Some(buffer::Event::ChatHistoryBeforeRequest) => {
+                                if let Some(buffer) = pane.buffer.data() {
+                                    let server = buffer.server();
+
+                                    if clients.get_server_supports_chathistory(server) {
+                                        if let Some(target) = buffer.target() {
+                                            let message_reference_type = clients
+                                                .get_server_chathistory_message_reference_type(
```

Comment:
We can consider `infinite_scroll` within the `scroll_view` and conditionally return the event or not. Then we can define a single event variant `Event::RequestMoreChatHistory` that's used from both the button message & scroll to top message and dedupe all this code.